src/cache.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * @fileoverview this file provides methods handling dependency cache
+ */
+
+import { join } from 'path';
+import os from 'os';
+import * as cache from '@actions/cache';
+import * as core from '@actions/core';
+import * as glob from '@actions/glob';
+
+const STATE_CACHE_PRIMARY_KEY = 'cache-primary-key';
+const CACHE_MATCHED_KEY = 'cache-matched-key';
+
+interface PackageManager {
+  id: 'maven' | 'gradle';
+  /**
+   * Paths of the file that specify the files to cache.
+   */
+  path: string[];
+  pattern: string[];
+}
+const supportedPackageManager: PackageManager[] = [
+  {
+    id: 'maven',
+    path: [join(os.homedir(), '.m2', 'repository')],
+    // https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#java---maven
+    pattern: ['**/pom.xml']
+  },
+  {
+    id: 'gradle',
+    path: [join(os.homedir(), '.gradle', 'caches'), join(os.homedir(), '.gradle', 'wrapper')],
+    // https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#java---gradle
+    pattern: ['**/*.gradle*', '**/gradle-wrapper.properties']
+  }
+];
+
+function findPackageManager(id: string): PackageManager {
+  const packageManager = supportedPackageManager.find(packageManager => packageManager.id === id);
+  if (packageManager === undefined) {
+    throw new Error(`unknown package manager specified: ${id}`);
+  }
+  return packageManager;
+}
+
+/**
+ * A function that generates a cache key to use.
+ * Format of the generated key will be "${{ platform }}-${{ id }}-${{ fileHash }}"".
+ * If there is no file matched to {@link PackageManager.path}, the generated key ends with a dash (-).
+ * @see {@link https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key|spec of cache key}
+ */
+async function computeCacheKey(packageManager: PackageManager) {
+  return `${process.env['RUNNER_OS']}-${packageManager.id}-${await glob.hashFiles(
```

Comment:
Could you please add to the cache key `setup-java` to look something like this: `setup-java-${process.env['RUNNER_OS']}-${packageManager.id}-${await glob.hashFiles(packageManager.pattern.join('\n'))}`

Nitpick: Could you please split this commands on the lines.